README.md
# Snake

This is a snake game

* You can move either with WASD or the arrow keys
* You need the following librarys to run the game:
    1. numpy
    2. scipy
    3. matplotlib
    4. pyopengl
    5. glfw
    6. ipython
    7. pillow
    8. jupyter

* The game runs with the command "python snake.py N" where N is the grid size
* Have fun :)

ChangeLog.md
Version 2.0
-----------

``RC.1`` Implemented game over screen
``B.2`` Implemented adaptative game speed to the grid size
``B.1`` Added textures to the snake head and body


Version 1.0
-----------
- ``RC.1`` Added the game score
- ``B.6`` Implemented the main logic of the game
- ``B.5`` Implemented when an apple is eaten it spawns in a random location
- ``B.4`` Implemented eat apples
- ``B.3`` Implemented collision with the borders
- ``B.2`` Updated the apple model and added the grid pattern
- ``B.1`` Implemented the basic apple and snake draw


view.py
import glfw
from OpenGL.GL import *
import sys
from model import *
from controller import Controller

N = int(sys.argv[1])

speed = 2/N

if __name__ == '__main__':

    # Initialize glfw
    if not glfw.init():
        sys.exit()

    width = 1000
    height = 1000

    window = glfw.create_window(width, height, 'Snake', None, None)

    if not window:
        glfw.terminate()
        sys.exit()

    glfw.make_context_current(window)

    controller = Controller()

    # Connecting the callback function 'on_key' to handle keyboard events
    glfw.set_key_callback(window, controller.on_key)

    # Assembling the shader program (pipeline) with both shaders
    pipeline = es.SimpleTransformShaderProgram()

    # Assembling the shader program (pipeline2) with both shaders
    pipeline2 = es.SimpleTextureTransformShaderProgram()

    glClearColor(16/255, 74/255, 3/255, 1.0)

    # Our shapes here are always fully painted
    glPolygonMode(GL_FRONT_AND_BACK, GL_FILL)

    snake = Snake()

    controller.set_model(snake)

    apple = Apple(N)

    border = Scene()

    logic = Logic(snake)

    end = EndScreeen()

    t0 = 0

    glEnable(GL_BLEND)
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)

    while not glfw.window_should_close(window):  # Dibujando --> 1. obtener el input
        
        ti = glfw.get_time()
        dt = ti - t0
        
        # Using GLFW to check for input events
        glfw.poll_events()  # OBTIENE EL INPUT --> CONTROLADOR --> MODELOS

        # Clearing the screen in both, color and depth
        glClear(GL_COLOR_BUFFER_BIT)
        # DIBUJAR LOS MODELOS
        glUseProgram(pipeline.shaderProgram)
        border.draw(pipeline, N)
        apple.draw(pipeline, N)
        glUseProgram(pipeline2.shaderProgram)
        snake.draw(pipeline2, N)
        logic.draw(pipeline2, N)

        if dt >= speed and snake.on:
            logic.movement(N)
            logic.borderCollision(N)
            logic.bodyCollision(N)
            logic.appleEaten(N, apple)
            t0 = ti

        if snake.on == False:
            end.draw(pipeline2, N)
        

        # Once the render is done, buffers are swapped, showing only the complete scene.
        glfw.swap_buffers(window)
        

    glfw.terminate()

controller.py
from model import *
import glfw
import sys


SIZE_IN_BYTES = 4


# A class to control the application
class Controller(object):
    def __init__(self):
        self.model = None
        

    def set_model(self, m):
        self.model = m

# we will use the global controller as communication with the callback function
    def on_key(self,window, key, scancode, action, mods):
        if action == glfw.PRESS:
            if key == glfw.KEY_ESCAPE:
                sys.exit()
            elif (key == glfw.KEY_UP or key == glfw.KEY_W) and self.model.direction != "DOWN":
                self.model.direction = "UP"
            elif (key == glfw.KEY_DOWN or key == glfw.KEY_S) and self.model.direction != "UP":
                self.model.direction = "DOWN"
            elif (key == glfw.KEY_LEFT or key == glfw.KEY_A) and self.model.direction != "RIGHT":
                self.model.direction = "LEFT"
            elif (key == glfw.KEY_RIGHT or key == glfw.KEY_D) and self.model.direction != "LEFT":
                self.model.direction = "RIGHT"


                








basic_shapes.py
# coding=utf-8
"""
Daniel Calderon, CC3501, 2019-2
vertices and indices for simple shapes
"""


# A simple class container to store vertices and indices that define a shape
class Shape:
    def __init__(self, vertices, indices, textureFileName=None):
        self.vertices = vertices
        self.indices = indices
        self.textureFileName = textureFileName


def createAxis(length=1.0):
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #    positions        colors
        -length, 0.0, 0.0, 0.0, 0.0, 0.0,
        length, 0.0, 0.0, 1.0, 0.0, 0.0,

        0.0, -length, 0.0, 0.0, 0.0, 0.0,
        0.0, length, 0.0, 0.0, 1.0, 0.0,

        0.0, 0.0, -length, 0.0, 0.0, 0.0,
        0.0, 0.0, length, 0.0, 0.0, 1.0]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1,
        2, 3,
        4, 5]

    return Shape(vertices, indices)


def createRainbowTriangle():
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #   positions        colors
        -0.5, -0.5, 0.0, 1.0, 0.0, 0.0,
        0.5, -0.5, 0.0, 0.0, 1.0, 0.0,
        0.0, 0.5, 0.0, 0.0, 0.0, 1.0]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [0, 1, 2]

    return Shape(vertices, indices)


def createRainbowQuad():
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #   positions        colors
        -0.5, -0.5, 0.0, 1.0, 0.0, 0.0,
        0.5, -0.5, 0.0, 0.0, 1.0, 0.0,
        0.5, 0.5, 0.0, 0.0, 0.0, 1.0,
        -0.5, 0.5, 0.0, 1.0, 1.0, 1.0]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2,
        2, 3, 0]

    return Shape(vertices, indices)


def createColorQuad(r, g, b):
    # Defining locations and colors for each vertex of the shape
    vertices = [
        #   positions        colors
        -0.5, -0.5, 0.0, r, g, b,
        0.5, -0.5, 0.0, r, g, b,
        0.5, 0.5, 0.0, r, g, b,
        -0.5, 0.5, 0.0, r, g, b]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2,
        2, 3, 0]

    return Shape(vertices, indices)

def createColorDegradeQuad(r, g, b, r2, g2, b2):
    # Defining locations and colors for each vertex of the shape
    vertices = [
        #   positions        colors
        -0.5, -0.5, 0.0, r, g, b,
        0.5, -0.5, 0.0, r, g, b,
        0.5, 0.5, 0.0, r, g, b,
        -0.5, 0.5, 0.0, r2, g2, b2]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2,
        2, 3, 0]

    return Shape(vertices, indices)


def createTextureQuad(image_filename, nx=1, ny=1):
    # Defining locations and texture coordinates for each vertex of the shape
    vertices = [
        #   positions        texture
        -0.5, -0.5, 0.0, 0, ny,
        0.5, -0.5, 0.0, nx, ny,
        0.5, 0.5, 0.0, nx, 0,
        -0.5, 0.5, 0.0, 0, 0]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2,
        2, 3, 0]

    textureFileName = image_filename

    return Shape(vertices, indices, textureFileName)


def createRainbowCube():
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #    positions         colors
        -0.5, -0.5, 0.5, 1.0, 0.0, 0.0,
        0.5, -0.5, 0.5, 0.0, 1.0, 0.0,
        0.5, 0.5, 0.5, 0.0, 0.0, 1.0,
        -0.5, 0.5, 0.5, 1.0, 1.0, 1.0,

        -0.5, -0.5, -0.5, 1.0, 1.0, 0.0,
        0.5, -0.5, -0.5, 0.0, 1.0, 1.0,
        0.5, 0.5, -0.5, 1.0, 0.0, 1.0,
        -0.5, 0.5, -0.5, 1.0, 1.0, 1.0]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2, 2, 3, 0,
        4, 5, 6, 6, 7, 4,
        4, 5, 1, 1, 0, 4,
        6, 7, 3, 3, 2, 6,
        5, 6, 2, 2, 1, 5,
        7, 4, 0, 0, 3, 7]

    return Shape(vertices, indices)


def createColorCube(r, g, b):
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #    positions        colors
        -0.5, -0.5, 0.5, r, g, b,
        0.5, -0.5, 0.5, r, g, b,
        0.5, 0.5, 0.5, r, g, b,
        -0.5, 0.5, 0.5, r, g, b,

        -0.5, -0.5, -0.5, r, g, b,
        0.5, -0.5, -0.5, r, g, b,
        0.5, 0.5, -0.5, r, g, b,
        -0.5, 0.5, -0.5, r, g, b]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2, 2, 3, 0,
        4, 5, 6, 6, 7, 4,
        4, 5, 1, 1, 0, 4,
        6, 7, 3, 3, 2, 6,
        5, 6, 2, 2, 1, 5,
        7, 4, 0, 0, 3, 7]

    return Shape(vertices, indices)


def createTextureCube(image_filename):
    # Defining locations and texture coordinates for each vertex of the shape
    vertices = [
        #   positions         texture coordinates
        # Z+
        -0.5, -0.5, 0.5, 0, 1,
        0.5, -0.5, 0.5, 1, 1,
        0.5, 0.5, 0.5, 1, 0,
        -0.5, 0.5, 0.5, 0, 0,

        # Z-
        -0.5, -0.5, -0.5, 0, 1,
        0.5, -0.5, -0.5, 1, 1,
        0.5, 0.5, -0.5, 1, 0,
        -0.5, 0.5, -0.5, 0, 0,

        # X+
        0.5, -0.5, -0.5, 0, 1,
        0.5, 0.5, -0.5, 1, 1,
        0.5, 0.5, 0.5, 1, 0,
        0.5, -0.5, 0.5, 0, 0
        ,

        # X-
        -0.5, -0.5, -0.5, 0, 1,
        -0.5, 0.5, -0.5, 1, 1,
        -0.5, 0.5, 0.5, 1, 0,
        -0.5, -0.5, 0.5, 0, 0,

        # Y+
        -0.5, 0.5, -0.5, 0, 1,
        0.5, 0.5, -0.5, 1, 1,
        0.5, 0.5, 0.5, 1, 0,
        -0.5, 0.5, 0.5, 0, 0,

        # Y-
        -0.5, -0.5, -0.5, 0, 1,
        0.5, -0.5, -0.5, 1, 1,
        0.5, -0.5, 0.5, 1, 0,
        -0.5, -0.5, 0.5, 0, 0
    ]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2, 2, 3, 0,  # Z+
        7, 6, 5, 5, 4, 7,  # Z-
        8, 9, 10, 10, 11, 8,  # X+
        15, 14, 13, 13, 12, 15,  # X-
        19, 18, 17, 17, 16, 19,  # Y+
        20, 21, 22, 22, 23, 20]  # Y-

    return Shape(vertices, indices, image_filename)


def createRainbowNormalsCube():
    sq3 = 0.57735027

    # Defining the location and colors of each vertex  of the shape
    vertices = [
        -0.5, -0.5, 0.5, 1.0, 0.0, 0.0, -sq3, -sq3, sq3,
        0.5, -0.5, 0.5, 0.0, 1.0, 0.0, sq3, -sq3, sq3,
        0.5, 0.5, 0.5, 0.0, 0.0, 1.0, sq3, sq3, sq3,
        -0.5, 0.5, 0.5, 1.0, 1.0, 1.0, -sq3, sq3, sq3,

        -0.5, -0.5, -0.5, 1.0, 1.0, 0.0, -sq3, -sq3, -sq3,
        0.5, -0.5, -0.5, 0.0, 1.0, 1.0, sq3, -sq3, -sq3,
        0.5, 0.5, -0.5, 1.0, 0.0, 1.0, sq3, sq3, -sq3,
        -0.5, 0.5, -0.5, 1.0, 1.0, 1.0, -sq3, sq3, -sq3]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [0, 1, 2, 2, 3, 0,
               4, 5, 6, 6, 7, 4,
               4, 5, 1, 1, 0, 4,
               6, 7, 3, 3, 2, 6,
               5, 6, 2, 2, 1, 5,
               7, 4, 0, 0, 3, 7]

    return Shape(vertices, indices)


def createColorNormalsCube(r, g, b):
    # Defining the location and colors of each vertex  of the shape
    vertices = [
        #   positions         colors   normals
        # Z+
        -0.5, -0.5, 0.5, r, g, b, 0, 0, 1,
        0.5, -0.5, 0.5, r, g, b, 0, 0, 1,
        0.5, 0.5, 0.5, r, g, b, 0, 0, 1,
        -0.5, 0.5, 0.5, r, g, b, 0, 0, 1,

        # Z-
        -0.5, -0.5, -0.5, r, g, b, 0, 0, -1,
        0.5, -0.5, -0.5, r, g, b, 0, 0, -1,
        0.5, 0.5, -0.5, r, g, b, 0, 0, -1,
        -0.5, 0.5, -0.5, r, g, b, 0, 0, -1,

        # X+
        0.5, -0.5, -0.5, r, g, b, 1, 0, 0,
        0.5, 0.5, -0.5, r, g, b, 1, 0, 0,
        0.5, 0.5, 0.5, r, g, b, 1, 0, 0,
        0.5, -0.5, 0.5, r, g, b, 1, 0, 0,

        # X-
        -0.5, -0.5, -0.5, r, g, b, -1, 0, 0,
        -0.5, 0.5, -0.5, r, g, b, -1, 0, 0,
        -0.5, 0.5, 0.5, r, g, b, -1, 0, 0,
        -0.5, -0.5, 0.5, r, g, b, -1, 0, 0,

        # Y+
        -0.5, 0.5, -0.5, r, g, b, 0, 1, 0,
        0.5, 0.5, -0.5, r, g, b, 0, 1, 0,
        0.5, 0.5, 0.5, r, g, b, 0, 1, 0,
        -0.5, 0.5, 0.5, r, g, b, 0, 1, 0,

        # Y-
        -0.5, -0.5, -0.5, r, g, b, 0, -1, 0,
        0.5, -0.5, -0.5, r, g, b, 0, -1, 0,
        0.5, -0.5, 0.5, r, g, b, 0, -1, 0,
        -0.5, -0.5, 0.5, r, g, b, 0, -1, 0
    ]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2, 2, 3, 0,  # Z+
        7, 6, 5, 5, 4, 7,  # Z-
        8, 9, 10, 10, 11, 8,  # X+
        15, 14, 13, 13, 12, 15,  # X-
        19, 18, 17, 17, 16, 19,  # Y+
        20, 21, 22, 22, 23, 20]  # Y-

    return Shape(vertices, indices)


def createTextureNormalsCube(image_filename):
    # Defining locations,texture coordinates and normals for each vertex of the shape
    vertices = [
        #   positions            tex coords   normals
        # Z+
        -0.5, -0.5, 0.5, 0, 1, 0, 0, 1,
        0.5, -0.5, 0.5, 1, 1, 0, 0, 1,
        0.5, 0.5, 0.5, 1, 0, 0, 0, 1,
        -0.5, 0.5, 0.5, 0, 0, 0, 0, 1,
        # Z-
        -0.5, -0.5, -0.5, 0, 1, 0, 0, -1,
        0.5, -0.5, -0.5, 1, 1, 0, 0, -1,
        0.5, 0.5, -0.5, 1, 0, 0, 0, -1,
        -0.5, 0.5, -0.5, 0, 0, 0, 0, -1,

        # X+
        0.5, -0.5, -0.5, 0, 1, 1, 0, 0,
        0.5, 0.5, -0.5, 1, 1, 1, 0, 0,
        0.5, 0.5, 0.5, 1, 0, 1, 0, 0,
        0.5, -0.5, 0.5, 0, 0, 1, 0, 0,
        # X-
        -0.5, -0.5, -0.5, 0, 1, -1, 0, 0,
        -0.5, 0.5, -0.5, 1, 1, -1, 0, 0,
        -0.5, 0.5, 0.5, 1, 0, -1, 0, 0,
        -0.5, -0.5, 0.5, 0, 0, -1, 0, 0,
        # Y+
        -0.5, 0.5, -0.5, 0, 1, 0, 1, 0,
        0.5, 0.5, -0.5, 1, 1, 0, 1, 0,
        0.5, 0.5, 0.5, 1, 0, 0, 1, 0,
        -0.5, 0.5, 0.5, 0, 0, 0, 1, 0,
        # Y-
        -0.5, -0.5, -0.5, 0, 1, 0, -1, 0,
        0.5, -0.5, -0.5, 1, 1, 0, -1, 0,
        0.5, -0.5, 0.5, 1, 0, 0, -1, 0,
        -0.5, -0.5, 0.5, 0, 0, 0, -1, 0
    ]

    # Defining connections among vertices
    # We have a triangle every 3 indices specified
    indices = [
        0, 1, 2, 2, 3, 0,  # Z+
        7, 6, 5, 5, 4, 7,  # Z-
        8, 9, 10, 10, 11, 8,  # X+
        15, 14, 13, 13, 12, 15,  # X-
        19, 18, 17, 17, 16, 19,  # Y+
        20, 21, 22, 22, 23, 20]  # Y-

    return Shape(vertices, indices, image_filename)


model.py
import numpy as np

import transformations as tr
import easy_shaders as es
import scene_graph as sg
import basic_shapes as bs
from OpenGL.GL import *

class Snake (object):
    """
    Creates the snake's head, and implements
    the head movement logic.
    """

    def __init__ (self):
        gpu_head_quad = es.toGPUShape(
            bs.createTextureQuad("img/Snake_Head.png"), GL_REPEAT, GL_NEAREST)
        head = sg.SceneGraphNode("head")
        head.transform = tr.translate(0,0,0)
        head.childs += [gpu_head_quad]
        self.posx = 0 
        self.posy = 0
        self.direction = "UP"
        self.model = head
        self.on = True
        
    def draw (self, pipeline, N):
        self.model.transform = np.matmul(
            tr.translate(self.posx,self.posy,0), tr.uniformScale(1/N))
        sg.drawSceneGraphNode(self.model, pipeline, 'transform')

    def update(self, N):
        if self.direction == "UP":
            self.posy += 1/N
        elif self.direction == "DOWN":
            self.posy -= 1/N
        elif self.direction == "LEFT":
            self.posx -= 1/N
        elif self.direction == "RIGHT":
            self.posx += 1/N

class Body(object):
    """
    Draws and implements the body of the snake.
    """
    def __init__ (self, x, y):
        gpu_body_quad = es.toGPUShape(
            bs.createTextureQuad("img/Snake_Body.png"), GL_REPEAT, GL_NEAREST)
        body = sg.SceneGraphNode("body")
        body.transform = tr.translate(0,0,0)
        body.childs += [gpu_body_quad]
        self.posx = x
        self.posy = y
        self.model = body
    
    def draw(self, pipeline, N):
        self.model.transform = tr.matmul([
            tr.translate(self.posx, self.posy, 0),
            tr.uniformScale( 1/N )])
        sg.drawSceneGraphNode(self.model, pipeline, 'transform')


class Logic(object):
    """
    This holds the main logic of the game, the movement
    of the body logic, the adding of a body logic, the collisions logics
    and the apple logics
    """
    def __init__ (self, snake):
        self.head = snake
        self.body = []

    def addBody (self):
        """
        Appends a body part on the end
        """
        if len(self.body) == 0:
            self.body.append(Body(self.head.posx, self.head.posy))
        else:
            self.body.append(Body(self.body[len(self.body)-1].posx, self.body[len(self.body)-1].posy))

    def movement (self, N):
        """
        Movement of the body
        """
        x_head = self.head.posx
        y_head = self.head.posy
        self.head.update(N)

        for i in range(len(self.body)):
            x_copy = self.body[i].posx
            y_copy = self.body[i].posy
            self.body[i].posx = x_head
            self.body[i].posy = y_head
            x_head = x_copy
            y_head = y_copy
        
    def bodyCollision(self, N):
        """
        Checks if the snake is colliding with his body
        """
        x_head = self.head.posx
        y_head = self.head.posy
        for i in self.body:
            if ((i.posx - 1/(2*N) < x_head < i.posx +1/(2*N)) and
                (i.posy - 1/(2*N) < y_head < i.posy +1/(2*N))):
                self.head.on = False
                print("Game Over, tu puntaje es:", len(self.body))
    
    def borderCollision (self, N):
        """
        Checks if the snake is colliding with the border
        """

        x_head = self.head.posx
        y_head = self.head.posy
        if x_head >= 1 - (1/(2*N)) or y_head >= 1 - (1/(2*N)):
            self.head.on = False
            print("Game Over, tu puntaje es:", len(self.body))
        elif x_head <= -1 + (1/(2*N)) or y_head <= -1 + (1/(2*N)):
            self.head.on = False
            print("Game Over, tu puntaje es:", len(self.body))

    def appleEaten (self, N, apple):
        """
        First we check if the apple is goin to be eaten by the snake,
        then we a start a cycle where generates 2 random numbers and
        those numbers will be the candidate coordinates, the we check
        if the apple is goin to be in a empty space, if is not going on
        a empty space, we generate 2 new numbers.
        """
        x_head = self.head.posx
        y_head = self.head.posy
        if ((apple.posx - 1/(2*N) < x_head < apple.posx +1/(2*N)) and 
            (apple.posy - 1/(2*N) < y_head < apple.posy +1/(2*N))):

            self.addBody()

            

            while True:

                aux = 0 #Helps to see if the apple is goin to go on a empty space
                x = np.random.randint(-10 , 10)
                y = np.random.randint(-10 , 10)
                for i in self.body:
                    if x < 0 and y < 0:
                        apple.posx = x/10 + (1/(N))
                        apple.posy = y/10 + (1/(N))
                        if not ((apple.posx - 1/N < i.posx < apple.posx +1/N) and 
                            (apple.posy - 1/N < i.posy < apple.posy +1/N)): 
                            aux += 1
                        else:
                            break

                    elif x >= 0 and y < 0:
                        apple.posx = x/10 - (1/(N))
                        apple.posy = y/10 + (1/(N))
                        if not ((apple.posx - 1/N < i.posx < apple.posx +1/N) and 
                            (apple.posy - 1/N < i.posy < apple.posy +1/N)): 
                            aux += 1
                        else:
                            break
                    elif x >= 0 and y >= 0:
                        apple.posx = x/10 - (1/(N))
                        apple.posy = y/10 - (1/(N))
                        if not ((apple.posx - 1/N < i.posx < apple.posx +1/N) and 
                            (apple.posy - 1/N < i.posy < apple.posy +1/N)): 
                            aux += 1
                        else:
                            break
                    elif x < 0 and y >= 0:
                        apple.posx = x/10 + (1/(N))
                        apple.posy = y/10 - (1/(N))
                        if not ((apple.posx - 1/N < i.posx < apple.posx +1/N) and 
                            (apple.posy - 1/N < i.posy < apple.posy +1/N)): 
                            aux += 1
                        else:
                            break
                if aux == len(self.body):
                    break
                
            #print("Comiste ñam")

    def draw(self, pipeline, N):
        for x in self.body:
            x.draw(pipeline, N)


class Scene (object):
    """
    Creates the borders of the map
    """
    def __init__ (self):
        gpu_head_quad = es.toGPUShape(
            bs.createColorQuad(40/255, 186/255, 35/255))
        border = sg.SceneGraphNode("border")
        border.transform = tr.translate(0,0,0)
        border.childs += [gpu_head_quad]
        transform_border = sg.SceneGraphNode('borderTR')
        transform_border.childs += [border]

        self.model = transform_border
        self.posx = 0
        self.posy = 0

    def draw (self, pipeline, N):
        self.model.transform = tr.scale(2 - (1 / N), 2 - (1 / N), 1)
        sg.drawSceneGraphNode(self.model, pipeline, 'transform')





class Apple (object):
    """
    Creates the apple shape
    """
    def __init__ (self,N):
        gpu_apple_quad = es.toGPUShape(
            bs.createColorDegradeQuad(217/255, 22/255, 22/255,
             217/255, 100/255, 150/255))
        apple = sg.SceneGraphNode("apple")
        apple.transform = tr.translate(0,0,0)
        apple.childs += [gpu_apple_quad]

        gpu_stick_quad = es.toGPUShape(
            bs.createColorQuad(69/255, 29/255, 4/255))
        stick = sg.SceneGraphNode("stick")
        stick.transform = tr.matmul([
            tr.translate(0,0.6,0), tr.scale(0.15, 0.30, 1)])
        stick.childs += [gpu_stick_quad]

        gpu_leaf_quad = es.toGPUShape(
            bs.createColorQuad(38/255, 128/255, 33/255))
        leaf = sg.SceneGraphNode("leaf")
        leaf.transform = tr.matmul([
            tr.translate(0.20,0.6,0),
             tr.scale(0.25, 0.15, 1), tr.rotationZ(0.5)])
        leaf.childs += [gpu_leaf_quad]

        apple_stick_leaf = sg.SceneGraphNode("apple_stick_leaf")
        apple_stick_leaf.childs += [apple, stick, leaf]
        transform_apple = sg.SceneGraphNode('appleTR')
        transform_apple.childs += [apple_stick_leaf]

        self.model = transform_apple
        self.posx = 0
        self.posy = 1/N

    def draw (self, pipeline, N):
        self.model.transform = tr.matmul([
            tr.translate(self.posx,self.posy,0), tr.uniformScale(1/N)])
        sg.drawSceneGraphNode(self.model, pipeline, 'transform')

class EndScreeen(object):

    """
    Displays the game over texture
    """

    def __init__(self):
        gpu_end_quad = es.toGPUShape(
                bs.createTextureQuad("img/GO.png"), GL_REPEAT, GL_NEAREST)
        end = sg.SceneGraphNode("end")
        end.transform = tr.translate(0,0,0)
        end.childs += [gpu_end_quad]
        self.posx = 0 
        self.posy = 0
        self.model = end
        
    def draw (self, pipeline, N):
        self.model.transform = np.matmul(
            tr.translate(self.posx+0.1,self.posy+0.05, 0), tr.rotationZ(np.pi/6))
        sg.drawSceneGraphNode(self.model, pipeline, 'transform')